File: spicedb-python/spicedb/consistency.py
"""Consistency strategy constructors for SpiceDB operations."""

from __future__ import annotations

from dataclasses import dataclass

from authzed.api.v1 import permission_service_pb2
from authzed.api.v1.core_pb2 import ZedToken


@dataclass(frozen=True)
class Consistency:
    """Opaque consistency strategy.

    Construct via ``full()``/``min_latency()``/``at_least()``/``snapshot()``/
    ``at_least_or_full()``/``at_least_or_min_latency()`` — never directly.
    """

    _proto: permission_service_pb2.Consistency

    def _to_proto(self) -> permission_service_pb2.Consistency:
        return self._proto


def full() -> Consistency:
    """Require fully consistent results."""
    return Consistency(_proto=permission_service_pb2.Consistency(fully_consistent=True))


def min_latency() -> Consistency:
    """Minimize latency, allowing slightly stale results."""
    return Consistency(_proto=permission_service_pb2.Consistency(minimize_latency=True))


def at_least(revision: str) -> Consistency:
    """Require results at least as fresh as the given revision."""
    return Consistency(
        _proto=permission_service_pb2.Consistency(
            at_least_as_fresh=ZedToken(token=revision)
        )
    )


def at_least_or_full(revision: str) -> Consistency:
    """Return at_least(revision) if revision is non-empty, otherwise full().

    Use this when you have an optional revision from a previous write and
    want the safest fallback.
    """
    if revision:
        return at_least(revision)
    return full()


def at_least_or_min_latency(revision: str) -> Consistency:
    """Return at_least(revision) if revision is non-empty, otherwise min_latency().

    Use this when you have an optional revision but prefer performance over
    full consistency as the fallback.
    """
    if revision:
        return at_least(revision)
    return min_latency()


def snapshot(revision: str) -> Consistency:
    """Require results from an exact snapshot revision."""
    return Consistency(
        _proto=permission_service_pb2.Consistency(
            at_exact_snapshot=ZedToken(token=revision)
        )
    )

File: spicedb-typescript/src/consistency.ts
import { create } from "@bufbuild/protobuf";
import {
  type Consistency as ProtoConsistency,
  ConsistencySchema,
} from "@spicedb/proto";
import { ZedTokenSchema } from "@spicedb/proto";

/**
 * Opaque consistency strategy. Construct via {@link full}, {@link minLatency},
 * {@link atLeast}, {@link snapshot}, {@link atLeastOrFull}, or
 * {@link atLeastOrMinLatency} -- never directly.
 */
export class Consistency {
  /** @internal */
  private constructor(private readonly proto: ProtoConsistency) {}

  /** @internal */
  static _wrap(proto: ProtoConsistency): Consistency {
    return new Consistency(proto);
  }

  /** @internal */
  _toProto(): ProtoConsistency {
    return this.proto;
  }
}

/**
 * Creates a consistency requirement for fully consistent reads.
 * This is the slowest option but guarantees the most up-to-date data.
 */
export function full(): Consistency {
  return Consistency._wrap(
    create(ConsistencySchema, {
      requirement: { case: "fullyConsistent", value: true },
    }),
  );
}

/**
 * Creates a consistency requirement that minimizes latency by using
 * the fastest snapshot available.
 */
export function minLatency(): Consistency {
  return Consistency._wrap(
    create(ConsistencySchema, {
      requirement: { case: "minimizeLatency", value: true },
    }),
  );
}

/**
 * Creates a consistency requirement that ensures data is at least as fresh
 * as the given revision token.
 *
 * @param revision - An opaque revision string returned from a previous write
 */
export function atLeast(revision: string): Consistency {
  return Consistency._wrap(
    create(ConsistencySchema, {
      requirement: {
        case: "atLeastAsFresh",
        value: create(ZedTokenSchema, { token: revision }),
      },
    }),
  );
}

/**
 * Returns atLeast(revision) if revision is non-empty, otherwise full().
 * Use this when you have an optional revision from a previous write and
 * want the safest fallback.
 *
 * @param revision - An opaque revision string, or empty string for full consistency
 */
export function atLeastOrFull(revision: string): Consistency {
  if (!revision) {
    return full();
  }
  return atLeast(revision);
}

/**
 * Returns atLeast(revision) if revision is non-empty, otherwise minLatency().
 * Use this when you have an optional revision but prefer performance over
 * full consistency as the fallback.
 *
 * @param revision - An opaque revision string, or empty string for min latency
 */
export function atLeastOrMinLatency(revision: string): Consistency {
  if (!revision) {
    return minLatency();
  }
  return atLeast(revision);
}

/**
 * Creates a consistency requirement that reads data at the exact given
 * snapshot. If the snapshot is no longer available, an error is returned.
 *
 * @param revision - An opaque revision string returned from a previous write
 */
export function snapshot(revision: string): Consistency {
  return Consistency._wrap(
    create(ConsistencySchema, {
      requirement: {
        case: "atExactSnapshot",
        value: create(ZedTokenSchema, { token: revision }),
      },
    }),
  );
}
